TWLight/users/authorization.py

```diff
@@ -481,38 +479,47 @@ def get(self, request, *args, **kwargs):
-                    return_url = next[0].encode("ascii", "ignore")
-                    # If there is anything left in the QueryDict after popping
-                    # 'next', append it to the return url. This preserves state
-                    # for filtered lists and redirected form submissions like
-                    # the partner suggestion form.
-                    if query_dict:
-                        return_url += "?" + urlencode(query_dict)
-                    logger.info(
-                        "User authenticated. Sending them on for "
-                        'post-login redirection per "next" parameter.'
-                    )
-                except KeyError:
-                    return_url = reverse_lazy(
-                        "users:editor_detail", kwargs={"pk": user.editor.pk}
-                    )
-                    logger.warning(
-                        'User authenticated. No "next" parameter '
-                        "for post-login redirection."
+                if user.userprofile.terms_of_use:
+                    # Translators: This message is shown when a user logs back in to the site after their first time.
+                    messages.add_message(request, messages.INFO, _("Welcome back!"))
+                    try:
+                        # Create a QueryDict from the 'get' session dict.
+                        query_dict = QueryDict(
+                            urlencode(request.session["get"]), mutable=True
+                        )
+                        # Pop the 'next' parameter out of the QueryDict.
+                        next = query_dict.pop("next")
+                        # Set the return url to the value of 'next'. Basic.
+                        return_url = next[0].encode("ascii", "ignore")
+                        # If there is anything left in the QueryDict after popping
+                        # 'next', append it to the return url. This preserves state
+                        # for filtered lists and redirected form submissions like
+                        # the partner suggestion form.
+                        if query_dict:
+                            return_url += "?" + urlencode(query_dict)
+                        logger.info(
+                            "User authenticated. Sending them on for "
+                            'post-login redirection per "next" parameter.'
+                        )
+                    except KeyError:
+                        return_url = reverse_lazy(
+                            "users:editor_detail", kwargs={"pk": user.editor.pk}
+                        )
+                        logger.warning(
+                            'User authenticated. No "next" parameter '
+                            "for post-login redirection."
+                        )
+                else:
+                    # Translators: This message is shown when a user logs back in to the site after their first time and hasn't agreed to the terms of use.
```

Comment:
The user should be sent on to their original destination after agreeing to terms. It looks like you're not preserving that here

Reply:
per conversation TODO for later